### ISSUE TITLE
Bug in taint+regex+hash/arrays

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3279]
Migrated from [rt.perl.org#5475](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5475) (status was 'resolved')

Searchable as RT5475$

**Comment by p5pRT** [comment]
### From samuele@artemisnet.it



Hello\. It seems that in hashes and arrays\, 
if tainting is activated\, the "g" action in regexp does not works
correctly\. This is a snipped example\, "taintest\.pl"\, invoked with
"taintest\.pl 'area=S&#8203;::=A\{emit\(\`ls \-la\`\)\} A&#8203;::=b' 'req=b'"


\#\!/usr/bin/perl \-T
\# \-\-\- part 1 \-\-\-
push\(@&#8203;in\,@&#8203;ARGV\);
foreach $i \(0 \.\. $\#in\) \{
&nbsp;   \($key\, $val\) = split\(/=/\,$in\[$i\]\,2\);
&nbsp;   \# \<\.\.\.snip\.\.\.>
&nbsp;   $in\{$key\} = $val;
&nbsp;   \}
\# \-\-\- end of part 1 \-\-\-
\# if you use the following line in place of "part 1" above\, it works\!
\# \(in this case hash is not tainted\)
\# $in\{area\} = 'S&#8203;::=A\{emit\(\`ls \-la\`\)\} A&#8203;::=b';
\# \-\-\-\-
\# but here this b\.\.\.rd enters in an infinite loop if taint is on
while\( $in\{area\} =~ /\(\[A\-Z\]\)&#8203;::=\(\.\+?\)\(?=\(\[A\-Z\]&#8203;::=|$\)\)/sg \) \{
&nbsp;   $in\{$1\}="$2";
&nbsp;   \# print for debug
&nbsp;   print "$1 is equal to $2\\n"; \}


Just FYI\, the regexp explores user's grammar productions\, coming from a
\<textarea>\</textarea> tags\, e\.g\.

S&#8203;::=aA\{\#a=A1\.s\+1; S\.s=\#a; emit\("String is \#a chars long"\)\}
A&#8203;::=xA\{A\.s=A1\.s\+1\} | @&#8203;\{A\.s=0\}

I posted the problem on comp\.lang\.perl\.misc\, and with another user\, Joe
Schaefer\, we have analyzed and tried to understand this problem\.

Joe wrote also&#8203;:

I can reproduce it on linux 5\.005\_03 and 5\.6\.
I think the taint flag is causing the regexp to fail 
to set pos\(\) for tainted hash and array elements\.  Binding to 
a tainted scalar seems to work fine\, though\.

Try testing it with the following lines substituted for 
the while\(\.\.\.\) line above&#8203;:

OK&#8203;:
&nbsp;        $\_ = $in\{area\};
&nbsp;        while \( $\_ =~ /\(\[A\-Z\]\)&#8203;::=\(\.\+?\)\(?=\(\[A\-Z\]&#8203;::=|$\)\)/sg \) \{     


ALSO BAD&#8203;:
&nbsp;        $\_\[0\] = $in\{area\};
&nbsp;        while \( $\_\[0\] =~ /\(\[A\-Z\]\)&#8203;::=\(\.\+?\)\(?=\(\[A\-Z\]&#8203;::=|$\)\)/sg \) \{     


Bye

&nbsp;   Samuele Manfrin \- Pisa\, Italy
&nbsp;   samuele@&#8203;null\.net \(home\) \- samuele@&#8203;artemisnet\.it \(work\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.6.0:

Configured by samuele at Thu Nov  9 18:07:40 MET 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.0.32, archname=i486-linux
    uname='linux 486lenna 2.0.32 #23 fri apr 24 14:39:36 met dst 1998 i486 '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=2.7.2
    cppflags='-I/usr/local/include'
    ccflags ='-I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lndbm -lgdbm -ldbm -ldb -ldl -lm -lc
    libc=/lib/libc.so.5.3.12, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /usr/local/lib/perl5/5.6.0/i486-linux
    /usr/local/lib/perl5/5.6.0
    /usr/local/lib/perl5/site_perl/5.6.0/i486-linux
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/samuele
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/bin:/usr/bin:/usr/local/bin:/sbin:/lib:/etc:/usr/etc:/usr/sbin:/usr/bin/X11:/usr/games:/home/samuele/bin:/usr/local/pgsql/bin:.
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @xdg

Confirmed that this bug still exists on Perl 5\.11\.0 on Linux\.  Binding
the regex to a tainted hash value causes an infinite loop\.  Copying the
same value to a scalar and binding the regex to the scalar does not\.

**Comment by p5pRT** [comment]
### From @iabyn

Now fixed by commit fd69380d5d5b95ef16e2521cf4251b34ee0ce151
in branch davem/post\-5\.12\, which should be merged back into blead
once 5\.12 has been released\, and thus appear in 5\.13 onwards\.

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @iabyn

Also required these to fully fix&#8203;:

2d961f6deff76f15f1e5c200d92c2ea6139bcdd2
39cf747a86645fde6898cd6d09d351d50755c2fa
